core/sql/cli/Context.cpp

```diff
@@ -3850,6 +3856,48 @@ void ContextCli::createMxcmpSession()
   mxcmpSessionInUse_ = TRUE;
 }
 
+// ----------------------------------------------------------------------------
+// Method:  updateMxcmpSession
+//
+// Updates security attributes in child arkcmp
+//
+// Returns: 0 = succeeded; -1 = failed
+// ----------------------------------------------------------------------------
+Int32 ContextCli::updateMxcmpSession()
+{
+  // If no child arkcmp, just return
+  if (getArkcmp()->getServer() == NULL)
+    return 0;
+
+  // Send changed user information to arkcmp process
+  CmpContext *cmpCntxt = CmpCommon::context();
+  ex_assert(cmpCntxt, "No compiler context exists");
+  NABoolean authOn = cmpCntxt->isAuthorizationEnabled();
+
+  // The message contains the following:
+  //   (auth state and user ID are delimited by commas)
+  //     authorization state (0 - off, 1 - on)
+  //     integer user ID
+  //     database user name
+  // See CmpStatement::process (CmpMessageDatabaseUser) for more details
+  Int32 userAsInt = (Int32) databaseUserID_;
+  char userMessage [MAX_AUTHID_AS_STRING_LEN + 1 + MAX_USERNAME_LEN + 1 + 2];
+  str_sprintf(userMessage, "%d,%d,%s", authOn, userAsInt, databaseUserName_);
+  char *pMessage = (char *)&userMessage;
+
+  // Send message to child arkcmp, if one exists
+  ExSqlComp::ReturnStatus cmpStatus;
+  ExSqlComp *exSqlComp = getArkcmp();
+  ex_assert(exSqlComp, "CliGlobals::getArkcmp() returned NULL");
+  cmpStatus = exSqlComp->sendRequest(EXSQLCOMP::DATABASE_USER,
+                     (const char *) pMessage,
+                     (ULng32) strlen(pMessage));
+  if (cmpStatus == ExSqlComp::ERROR)
+    return -1;
+
+  return 0;
+}
+
```

Comment:
Sounds good!

On Tue, Jun 7, 2016 at 9:47 AM, Roberta Marton <notifications@github.com>
wrote:

> In core/sql/cli/Context.cpp
> <https://github.com/apache/incubator-trafodion/pull/523#discussion_r66085087>
> :
>
> > +  str_sprintf(userMessage, "%d,%d,%s", authOn, userAsInt, databaseUserName_);
> > +  char *pMessage = (char *)&userMessage;
> > +
> > +  // Send message to child arkcmp, if one exists
> > +  ExSqlComp::ReturnStatus cmpStatus;
> > +  ExSqlComp *exSqlComp = getArkcmp();
> > +  ex_assert(exSqlComp, "CliGlobals::getArkcmp() returned NULL");
> > +  cmpStatus = exSqlComp->sendRequest(EXSQLCOMP::DATABASE_USER,
> > +                     (const char *) pMessage,
> > +                     (ULng32) strlen(pMessage));
> > +  if (cmpStatus == ExSqlComp::ERROR)
> > +    return -1;
> > +
> > +  return 0;
> > +}
> > +
>
> The code does propagate the change through the child, grandchild, etc. For
> drop authorization there are 4 generations of arkcmps and I need to
> propagate this value through them all - there is actually are regression
> test for this - udr/102.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/incubator-trafodion/pull/523/files/626370ee584eb5c3c847c01cd8615f204d29c3ff#r66085087>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AHuuAuR4LTvaAzHQ1131fgRGmjwnsmqvks5qJYSZgaJpZM4IubeB>
> .
>



-- 
Regards, --Qifan